### PR TITLE
vsock saga, episode 2: the connection state machine

### PR DESCRIPTION
**Pull request by dhrgit**
Issue #, if available: #650 

**Description of changes**

This PR adds a vsock connection state machine, that will later be used by the Unix Domain Sockets backend. This machine, residing under `devices::virtio::vsock::csm`, implements the bulk of the vsock communication protocol, and is designed to work, not necessarily with a backing Unix socket, but with any implementor of `Read + Write + AsRawFd`.

**Note**: this PR is built on top of #1106, so not all commits in this branch are in its scope.

--------
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: devices/src/virtio/vsock/csm/connection.rs [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+
+                // We might've just consumed some data. If that's the case, we might need to update
+                // the peer on our buffer space situation, so that it can keep sending data packets
+                // our way.
+                if self.peer_needs_credit_update() {
+                    self.pending_rx.insert(PendingRx::CreditUpdate);
+                }
+            }
+
+            // Next up: receing a response / confirmation for a host-initiated connection. We'll
```
Comment (alexandruag) [inline devices/src/virtio/vsock/csm/connection.rs]: receiving

File: devices/src/virtio/vsock/csm/connection.rs [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+            // move to an Established state, and pass on the good news through the host stream.
+            ConnState::LocalInit if pkt.op() == uapi::VSOCK_OP_RESPONSE => {
+                self.expiry = None;
+                self.send_bytes(format!("OK {}\n", self.local_port).as_bytes())
```
Comment (alexandruag) [inline devices/src/virtio/vsock/csm/connection.rs]: This is probably very very unlikely, but I don't think it's impossible to be in a situation where the next write does not manage to transmit the entire message here. I guess one option is to simply generate an error if we can't write them all in one go.

File: devices/src/virtio/vsock/csm/connection.rs [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+                });
+            self.fwd_cnt += Wrapping(flushed as u32);
+
+            // If we've freed up some more buffer space, we may need to let the peer know it can
+            // safely send more data our way.
+            if self.peer_needs_credit_update() {
+                self.pending_rx.insert(PendingRx::CreditUpdate);
+            }
+
+            // If this connection was shutting down, but is wating to drain the TX buffer
```
Comment (alexandruag) [inline devices/src/virtio/vsock/csm/connection.rs]: waiting

File: devices/src/virtio/vsock/csm/connection.rs [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+                // Any other connection state is invalid at this point, and we need to kill it with
+                // fire.
+                pkt.set_op(uapi::VSOCK_OP_RST);
+                return Ok(());
+            }
+        }
+
+        // Oh wait, before we start bringing in the big data, can our peer handle receiving so much
+        // bytey goodness?
+        if self.need_credit_update_from_peer() {
```
Comment (alexandruag) [inline devices/src/virtio/vsock/csm/connection.rs]: Didn't not think this through very well :D, but just wondering, is it possible to generate a lot of ops/traffic because we keep calling `recv_pkt` because there's pending data, but we don't know if the goodness can be handled, so we also keep sending these requests, until one finally comes back? So we no longer attempt to send anything if we've sent a credit request before receiving the associated response? Does it make sense to worry about this? :-s

File: devices/src/virtio/vsock/csm/connection.rs [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+                self.send_bytes(buf)?;
+            } else {
+                // Couldn't drain the TX buffer, so the new data needs to go on top of the buffer.
+                self.tx_buf.push(buf)?;
+            }
+            return Ok(());
+        }
+
+        // The TX buffer is empty, so we can try to write straight to the host stream.
+        let written = match self.stream.write(buf) {
```
Comment (alexandruag) [inline devices/src/virtio/vsock/csm/connection.rs]: I guess here goes the "write that returns 0" and `WouldBlock` chat again. In particular, if we keep acting like nothing really happened when we get 0 from write, or `WouldBlock` happens, is there any risk of generating a tight loop of such operations just because there's data to send, or we have logic which backs off and tries later somewhere?

File: devices/src/virtio/vsock/csm/connection.rs [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+                    );
+                    return Ok(());
+                }
+
+                // Unwrapping here is safe, since we just checked `pkt.buf()` above.
+                let buf_slice = &pkt.buf().unwrap()[..(pkt.len() as usize)];
+                if let Err(err) = self.send_bytes(buf_slice) {
+                    // If we can't write to the host stream, that's an unrecoverable error, so we'll
+                    // terminate this connection.
+                    info!(
```
Comment (dianpopa) [inline devices/src/virtio/vsock/csm/connection.rs]: Shouldn t this be an error since we are terminating the vsock connection at an unexpected time?

File: devices/src/virtio/vsock/csm/connection.rs [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+            // move to an Established state, and pass on the good news through the host stream.
+            ConnState::LocalInit if pkt.op() == uapi::VSOCK_OP_RESPONSE => {
+                self.expiry = None;
+                self.send_bytes(format!("OK {}\n", self.local_port).as_bytes())
+                    .and_then(|_| {
+                        self.state = ConnState::Established;
+                        Ok(())
+                    })
+                    .unwrap_or_else(|_| {
+                        self.kill();
```
Comment (dianpopa) [inline devices/src/virtio/vsock/csm/connection.rs]: Shouldn t we log an error in this case?

File: devices/src/virtio/vsock/csm/connection.rs [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+
+            // If we've freed up some more buffer space, we may need to let the peer know it can
+            // safely send more data our way.
+            if self.peer_needs_credit_update() {
+                self.pending_rx.insert(PendingRx::CreditUpdate);
+            }
+
+            // If this connection was shutting down, but is wating to drain the TX buffer
+            // before forcefull termination, the wait might be over.
+            if self.state == ConnState::PeerClosed(true, true) && self.tx_buf.is_empty() {
```
Comment (dianpopa) [inline devices/src/virtio/vsock/csm/connection.rs]: Is it possible to have the `will_not_send_anymore_data` flag on and be wating to drain the TX buffer `tx_buf`?

File: devices/src/virtio/vsock/csm/connection.rs [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+                        self.local_port, self.peer_port, err
+                    );
+                    self.kill();
+                    0
+                });
+            self.fwd_cnt += Wrapping(flushed as u32);
+
+            // If we've freed up some more buffer space, we may need to let the peer know it can
+            // safely send more data our way.
+            if self.peer_needs_credit_update() {
```
Comment (dianpopa) [inline devices/src/virtio/vsock/csm/connection.rs]: Is there a sense in sending a credit update if ulteriorly we send the RST flag (see the next if)? aren't these 2 ifs disjunctive?

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+use std::num::Wrapping;
+
+use super::defs;
+use super::{Error, Result};
+
+/// A simple ring-buffer implementation, used by vsock connections to buffer TX (guest -> host) data.
+/// Memory for this buffer is allocated lazily, since buffering will only be needed when the host
+/// can't read fast enough.
+///
+pub struct TxBuf {
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: I agree this could be turned in to a general purpose ring buffer, but there are few things to iron-out first:

- the `push()` behavior: push all or nothing; this check is currently made here, instead of at the buffer user's end, which might not be suitable for all use cases; it's fine here, since data needs to be stored fully, or the whole connection gets dropped, as insufficient buffer space implies a misbehaving peer;
- the lazy uninit memory allocation might, again, not be generally desirable;
- the non-configurable buffer size.

While I agree that all of the above can be implemented as user-configurable, my reasoning was that designing that flexible interface could be postponed until the moment we need another ring buffer somewhere. So, lazy allocation of coding resources, I guess? I'm not particularly fond of this approach, but it seemed like an acceptable compromise at the time.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    /// Ring-buffer head offset - where new data is pushed to.
+    head: Wrapping<u32>,
+    /// Ring-buffer tail offset - where data is flushed from.
+    tail: Wrapping<u32>,
+}
+
+impl TxBuf {
+    /// Total buffer size, in bytes.
+    ///
+    const SIZE: usize = defs::CONN_TX_BUF_SIZE;
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: Yes, that would have to change, if we make this a generic ring buffer. Also, we will want to switch from `mem::uninitialized` to `mem::MaybeUninit`.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+    pub fn len(&self) -> usize {
+        (self.head - self.tail).0 as usize
+    }
+
+    /// Push a byte slice onto the ring-buffer.
+    ///
+    /// Either the entire source slice will be pushed to the ring-buffer, or none of it, if there
+    /// isn't enough room, in which case `Err(Error::TxBufFull)` is returned.
+    ///
+    pub fn push(&mut self, src: &[u8]) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: It's for the fail-fast-if-we-can't-push-everything "feature". See my previous comment on generalizing the ring buffer.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        // - one write, if the tail doesn't need to wrap around to reach the head; or
+        // - two writes, if the tail would wrap around: tail to slice end, then slice end to head.
+
+        // First write length: the lesser of tail to slice end, or tail to head.
+        let len_to_write = std::cmp::min(Self::SIZE - tail_ofs, self.len());
+
+        // It's safe to unwrap here, since we've already checked if the buffer was empty.
+        let data = self.data.as_ref().unwrap();
+
+        // Issue the first write and absorb any `WouldBlock` error (we can just try again later).
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: I agree, to some extent. This was there due to some error in judgement I made, when designing the connection state machine. I.e. I would try to flush the TX buffer proactively (not waiting for an EPOLLOUT, even if I was registered to receive one), so `WouldBlock` could've been expected here. In retrospect, that wasn't the best decision, so I've changed that behavior now and also removed the absorption from here.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+        let written = match sink.write(&data[tail_ofs..(tail_ofs + len_to_write)]) {
+            Ok(written) => written,
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: It's removed now.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+mod tests {
+    use super::*;
+    use std::io::Error as IoError;
+    use std::io::Result as IoResult;
+    use std::io::Write;
+
+    struct TestSink {
+        data: Vec<u8>,
+        err: Option<IoError>,
+        capacity: usize,
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: Yes, the sink capacity can be dynamically adjusted to be less than `self.data.capacity`. `TestSink::capacity` is the total number of bytes that the sink can take in. So, to test an incomplete flush, `TestSink::capacity` is set to be less than `TestSink::data::capacity`, then set back to `TestSink::data::capacity` to test flushing the rest of the buffer. In retrospect, it is kind of an awkward testing pattern, I agree though.

File: devices/src/virtio/vsock/csm/txbuf.rs [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    impl TestSink {
+        const DEFAULT_CAPCITY: usize = 2 * TxBuf::SIZE;
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/txbuf.rs]: Fixed.

File: devices/src/virtio/vsock/csm/txbuf.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (dianpopa) [reply devices/src/virtio/vsock/csm/txbuf.rs]: I am ok with leaving it as it is.

File: devices/src/virtio/vsock/csm/connection.rs [hunk 17]
```diff
@@ -0,0 +1,632 @@
...
+            pkt.set_op(uapi::VSOCK_OP_CREDIT_UPDATE);
+            self.last_fwd_cnt_to_peer = self.fwd_cnt;
+            return Ok(());
+        }
+
+        // Alright, if we got to here, we need to cough up a data packet. We've already checked for
+        // all other pending RX indications.
+        self.pending_rx.remove(PendingRx::Rw);
+
+        match self.state {
```
[the remark this answers is not in the document]
Comment (dianpopa) [reply devices/src/virtio/vsock/csm/connection.rs]: I prefer the if but I agree it turns to a rather long condition. You can leave it as it is.

File: devices/src/virtio/vsock/csm/connection.rs [hunk 18]
```diff
@@ -0,0 +1,632 @@
...
+        // on it if we really have nothing else to say.
+        if self.pending_rx.remove(PendingRx::CreditUpdate) && !self.has_pending_rx() {
+            pkt.set_op(uapi::VSOCK_OP_CREDIT_UPDATE);
+            self.last_fwd_cnt_to_peer = self.fwd_cnt;
+            return Ok(());
+        }
+
+        // Alright, if we got to here, we need to cough up a data packet. We've already checked for
+        // all other pending RX indications.
+        self.pending_rx.remove(PendingRx::Rw);
```
[the remark this answers is not in the document]
Comment (dianpopa) [reply devices/src/virtio/vsock/csm/connection.rs]: I was not referring to calling `recv_pkt` being an invalid state. Let me rephrase it.
`self.pending_rx.remove` returns a bool (that you ignore), right? If the function were to return false, is there sense in going ahead and reading data?

[the remark this answers is not in the document]
Comment (dhrgit) [reply devices/src/virtio/vsock/csm/connection.rs]: Ah, right. I've gone through this section again and simplified some stuff. Now, a read is not attempted if `self.pending_rx` doesn't contain `PendingRx::Rw`. That's because reading (and writing, for that matter) data is fully driven by epoll. So, we shouldn't be attempting to read before `self.notify()` is called with `EPOLLIN` (which sets `Rw` inside `self.pendingRx`).

**Comment by dhrgit** [comment]
Closed as a subset of #1176 .
